File: lab4/c/bad5.c
// Giles Reger, 2019

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <limits.h>
#include <stdbool.h>

#include "pq.h"


struct llist {
  Value_Type value1;
  Value_Type value2;
  int priority1;
  int priority2;
};

struct llist* initialize_pq (int size){
  struct llist* list = malloc(sizeof(struct llist));
  list->value1=NULL;
  list->value2=NULL;
  return list;
}     

void tidy (struct llist* list){
}

bool contains(struct llist* pq, Value_Type value, int priority){
  return false;
}

void insert(struct llist* list, Value_Type value, int priority){
  if(list->value1==NULL){ list->value1=value; list->priority1=priority;}
  else{ 
    list->value2=value; list->priority2=priority;
    if(list->priority2 < list->priority1){
      Value_Type tv = list->value1;
      int tp = list->priority1;
      list->value1=list->value2;
      list->priority1=list->priority2;
      list->value2=tv;
      list->priority2=tp;
    }
  }
}

bool is_empty(struct llist* pq){ 
  return false;
}

Value_Type pop_min(struct llist* pq){
  Value_Type res = pq->value1;
  pq->value1=pq->value2;
  pq->priority1=pq->priority2;
  pq->value2=NULL;
  return res;
}

void print(struct llist* list){
}

File: lab4/python/bad5.py
import sys

class llist:
    def __init__(self, value1=None, value2=None, priority=0):
        self.value1 = value1
        self.value2 = value2
        self.priority1 = priority
        self.priority2 = priority
        self.next = None
        self.prev = None
        
    def contains(self, value, priority):
        return False
        
    def insert(self, value, priority):
        if (self.value1 == None):
            self.value1=value
            self.priority1=priority
        else:
            self.value2 = value
            self.priority2 = priority
            if (self.priority2 < self.priority1):
                tv = self.value1
                tp = self.priority1
                self.value1 = self.value2
                self.priority1 = self.priority2
                self.value2 = tv
                self.priority2 = tp
        
    def is_empty(self):
        return False
        
    def pop_min(self):
        res = self.value1
        self.value1 = self.value2
        self.priority1 = self.priority2
        self.value2 = None
        return res
        
    def print(self):
        print("No print implemented")
        
